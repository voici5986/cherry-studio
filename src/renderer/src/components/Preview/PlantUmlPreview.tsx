import { loggerService } from '@logger'
import pako from 'pako'
import React, { memo, useCallback, useEffect } from 'react'

import { useDebouncedRender } from './hooks/useDebouncedRender'
import ImagePreviewLayout from './ImagePreviewLayout'
import { BasicPreviewHandles, BasicPreviewProps } from './types'
import { renderSvgInShadowHost } from './utils'

const logger = loggerService.withContext('PlantUmlPreview')

const PlantUMLServer = 'https://www.plantuml.com/plantuml'
function encode64(data: Uint8Array) {
  let r = ''
  for (let i = 0; i < data.length; i += 3) {
    if (i + 2 === data.length) {
      r += append3bytes(data[i], data[i + 1], 0)
    } else if (i + 1 === data.length) {
      r += append3bytes(data[i], 0, 0)
    } else {
      r += append3bytes(data[i], data[i + 1], data[i + 2])
    }
  }
  return r
}

function encode6bit(b: number) {
  if (b < 10) {
    return String.fromCharCode(48 + b)
  }
  b -= 10
  if (b < 26) {
    return String.fromCharCode(65 + b)
  }
  b -= 26
  if (b < 26) {
    return String.fromCharCode(97 + b)
  }
  b -= 26
  if (b === 0) {
    return '-'
  }
  if (b === 1) {
    return '_'
  }
  return '?'
}

function append3bytes(b1: number, b2: number, b3: number) {
  const c1 = b1 >> 2
  const c2 = ((b1 & 0x3) << 4) | (b2 >> 4)
  const c3 = ((b2 & 0xf) << 2) | (b3 >> 6)
  const c4 = b3 & 0x3f
  let r = ''
  r += encode6bit(c1 & 0x3f)
  r += encode6bit(c2 & 0x3f)
  r += encode6bit(c3 & 0x3f)
  r += encode6bit(c4 & 0x3f)
  return r
}
/**
 * https://plantuml.com/zh/code-javascript-synchronous
 * To use PlantUML image generation, a text diagram description have to be :
    1. Encoded in UTF-8
    2. Compressed using Deflate algorithm
    3. Reencoded in ASCII using a transformation _close_ to base64
 */
function encodeDiagram(diagram: string): string {
  const utf8text = new TextEncoder().encode(diagram)
  const compressed = pako.deflateRaw(utf8text)
  return encode64(compressed)
}

function getPlantUMLImageUrl(format: 'png' | 'svg', diagram: string, isDark?: boolean) {
  const encodedDiagram = encodeDiagram(diagram)
  if (isDark) {
    return `${PlantUMLServer}/d${format}/${encodedDiagram}`
  }
  return `${PlantUMLServer}/${format}/${encodedDiagram}`
}

const PlantUmlPreview = ({
  children,
  enableToolbar = false,
  ref
}: BasicPreviewProps & { ref?: React.RefObject<BasicPreviewHandles | null> }) => {
  // 定义渲染函数
  const renderPlantUml = useCallback(async (content: string, container: HTMLDivElement) => {
    const url = getPlantUMLImageUrl('svg', content, false)
    const response = await fetch(url)
    if (!response.ok) {
      if (response.status === 400) {
        throw new Error(
          'Diagram rendering failed (400): This is likely due to a syntax error in the diagram. Please check your code.'
        )
      }
      if (response.status >= 500) {
        throw new Error(
          `Diagram rendering failed (${response.status}): The PlantUML server is temporarily unavailable. Please try again later.`
        )
      }
      throw new Error(`Diagram rendering failed, server returned: ${response.status} ${response.statusText}`)
    }

    const text = await response.text()
    renderSvgInShadowHost(text, container)
  }, [])

  // 使用预览渲染器 hook
  const { containerRef, error, isLoading } = useDebouncedRender(children, renderPlantUml, {
    debounceDelay: 300
  })

  // 记录网络错误
  useEffect(() => {
    if (error && error.includes('Failed to fetch')) {
      logger.warn('Network Error: Unable to connect to PlantUML server. Please check your network connection.')
    } else if (error) {
      logger.warn(error)
    }
  }, [error])

  return (
    <ImagePreviewLayout
      loading={isLoading}
      error={error}
      enableToolbar={enableToolbar}
      ref={ref}
      imageRef={containerRef}
      source="plantuml">
      <div ref={containerRef} className="plantuml-preview special-preview" />
    </ImagePreviewLayout>
  )
}

export default memo(PlantUmlPreview)
