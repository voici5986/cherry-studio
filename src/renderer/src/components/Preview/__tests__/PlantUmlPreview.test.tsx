import PlantUmlPreview from '@renderer/components/Preview/PlantUmlPreview'
import { render, screen } from '@testing-library/react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

// Use vi.hoisted to manage mocks
const mocks = vi.hoisted(() => ({
  ImagePreviewLayout: vi.fn(({ children, loading, error, enableToolbar, source }) => (
    <div data-testid="image-preview-layout" data-source={source}>
      {enableToolbar && <div data-testid="toolbar">Toolbar</div>}
      {loading && <div data-testid="loading">Loading...</div>}
      {error && <div data-testid="error">{error}</div>}
      <div data-testid="preview-content">{children}</div>
    </div>
  )),
  renderSvgInShadowHost: vi.fn(),
  useDebouncedRender: vi.fn(),
  logger: {
    warn: vi.fn()
  }
}))

vi.mock('@renderer/components/Preview/ImagePreviewLayout', () => ({
  default: mocks.ImagePreviewLayout
}))

vi.mock('@renderer/components/Preview/utils', () => ({
  renderSvgInShadowHost: mocks.renderSvgInShadowHost
}))

vi.mock('@renderer/components/Preview/hooks/useDebouncedRender', () => ({
  useDebouncedRender: mocks.useDebouncedRender
}))

describe('PlantUmlPreview', () => {
  const diagram = '@startuml\nA -> B\n@enduml'
  const mockContainerRef = { current: document.createElement('div') }

  // Helper function to create mock useDebouncedRender return value
  const createMockHookReturn = (overrides = {}) => ({
    containerRef: mockContainerRef,
    error: null,
    isLoading: false,
    triggerRender: vi.fn(),
    cancelRender: vi.fn(),
    clearError: vi.fn(),
    setLoading: vi.fn(),
    ...overrides
  })

  beforeEach(() => {
    // Setup default successful state
    mocks.useDebouncedRender.mockReturnValue(createMockHookReturn())
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<PlantUmlPreview enableToolbar>{diagram}</PlantUmlPreview>)
      expect(container).toMatchSnapshot()
    })

    it('should handle valid plantuml diagram', () => {
      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      // Component should render without throwing
      expect(screen.getByTestId('image-preview-layout')).toBeInTheDocument()
      expect(mocks.useDebouncedRender).toHaveBeenCalledWith(
        diagram,
        expect.any(Function),
        expect.objectContaining({ debounceDelay: 300 })
      )
    })

    it('should handle empty content', () => {
      render(<PlantUmlPreview>{''}</PlantUmlPreview>)

      // Component should render without throwing
      expect(screen.getByTestId('image-preview-layout')).toBeInTheDocument()
      expect(mocks.useDebouncedRender).toHaveBeenCalledWith('', expect.any(Function), expect.any(Object))
    })
  })

  describe('loading state', () => {
    it('should show loading indicator when rendering', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ isLoading: true }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      expect(screen.getByTestId('loading')).toBeInTheDocument()
    })

    it('should not show loading indicator when not rendering', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ isLoading: false }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      expect(screen.queryByTestId('loading')).not.toBeInTheDocument()
    })
  })

  describe('error handling', () => {
    it('should show network error message', () => {
      const networkError = 'Network Error: Unable to connect to PlantUML server. Please check your network connection.'
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: networkError }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      const errorElement = screen.getByTestId('error')
      expect(errorElement).toBeInTheDocument()
      expect(errorElement).toHaveTextContent(networkError)
    })

    it('should show syntax error message for invalid diagram', () => {
      const syntaxError =
        'Diagram rendering failed (400): This is likely due to a syntax error in the diagram. Please check your code.'
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: syntaxError }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      const errorElement = screen.getByTestId('error')
      expect(errorElement).toBeInTheDocument()
      expect(errorElement).toHaveTextContent(syntaxError)
    })

    it('should show server error message', () => {
      const serverError =
        'Diagram rendering failed (503): The PlantUML server is temporarily unavailable. Please try again later.'
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: serverError }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      const errorElement = screen.getByTestId('error')
      expect(errorElement).toBeInTheDocument()
      expect(errorElement).toHaveTextContent(serverError)
    })

    it('should show generic error message for other errors', () => {
      const genericError = "Diagram rendering failed, server returned: 418 I'm a teapot"
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: genericError }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      const errorElement = screen.getByTestId('error')
      expect(errorElement).toBeInTheDocument()
      expect(errorElement).toHaveTextContent(genericError)
    })

    it('should not show error when rendering is successful', () => {
      mocks.useDebouncedRender.mockReturnValue(createMockHookReturn({ error: null }))

      render(<PlantUmlPreview>{diagram}</PlantUmlPreview>)

      expect(screen.queryByTestId('error')).not.toBeInTheDocument()
    })
  })

  describe('ref forwarding', () => {
    it('should forward ref to ImagePreviewLayout', () => {
      const ref = { current: null }
      render(<PlantUmlPreview ref={ref}>{diagram}</PlantUmlPreview>)

      // The ref should be passed to ImagePreviewLayout
      expect(mocks.ImagePreviewLayout).toHaveBeenCalledWith(expect.objectContaining({ ref }), undefined)
    })
  })
})
