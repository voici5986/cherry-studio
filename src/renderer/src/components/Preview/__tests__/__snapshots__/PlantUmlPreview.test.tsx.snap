// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlantUmlPreview > basic rendering > should match snapshot 1`] = `
.c0 {
  --shadow-host-background-color: white;
  --shadow-host-border: 0.5px solid var(--color-code-background);
  --shadow-host-border-radius: 8px;
}

<div>
  <div
    data-source="plantuml"
    data-testid="image-preview-layout"
  >
    <div
      data-testid="toolbar"
    >
      Toolbar
    </div>
    <div
      data-testid="preview-content"
    >
      <div
        class="c0 plantuml-preview special-preview"
      />
    </div>
  </div>
</div>
`;
